### DISCUSSION TITLE
OutOfBoundsError: 0

### DISCUSSION THREAD

**Discussion by Xyx105** [Parcels-code/Parcels#1412] [General]
Hi,

I keep getting a "OutOfBoundsError: 0" error. I have tried the fix in #973 but  I still keep getting the same error.

Here is my code:
`filenames = {
    "U": {
        "lon": f"output.nc",
        "lat": f"output.nc",
        "data": f"NC Files/NEATL*.nc",
    },
    "V": {
        "lon": f"output.nc",
        "lat": f"output.nc",
        "data": f"NC Files/NEATL*.nc",
    },
}
             

variables = {'U': 'u',
             'V': 'v'}
             

dimensions = {"lon": "lon_u", "lat": "lat_u", "time": "ocean_time"}

indices = {'depth' : [40]}

fieldset = FieldSet.from_nemo(
    filenames, variables, dimensions, indices, allow_time_extrapolation=True,
)

#indices = {'depth' : [40]}

fieldset.computeTimeChunk(fieldset.U.grid.time[0], 6)
fieldset.computeTimeChunk(fieldset.V.grid.time[0], 6)

npart = 1
lonp = 4.7 
latp = 60 

pset = ParticleSet.from_list(fieldset, JITParticle, lon=lonp, lat=latp,  )

pfile = ParticleFile("nemo_particles", pset, outputdt=delta(days=5))
kernels = pset.Kernel(AdvectionRK4) 

pset.show(fieldset.U)

pset.execute(kernels, runtime=delta(days=1), dt=delta(hours=6), output_file=pfile)`

It produces this result
![Capture](https://github.com/OceanParcels/parcels/assets/127397110/d1d3184c-89f2-4032-9781-dd070b3009f0)

and I get this error message:
`---------------------------------------------------------------------------
OutOfBoundsError                          Traceback (most recent call last)
Cell In[16], line 82
     78 kernels = pset.Kernel(AdvectionRK4) 
     80 pset.show(fieldset.U)
---> 82 pset.execute(kernels, runtime=delta(days=1), dt=delta(hours=6), output_file=pfile)

File ~\Anaconda3\envs\parcels\Lib\site-packages\parcels\particleset\baseparticleset.py:548, in BaseParticleSet.execute(self, pyfunc, pyfunc_inter, endtime, runtime, dt, moviedt, recovery, output_file, movie_background_field, verbose_progress, postIterationCallbacks, callbackdt)
    546 # If we don't perform interaction, only execute the normal kernel efficiently.
    547 if self.interaction_kernel is None:
--> 548     self.kernel.execute(self, endtime=next_time, dt=dt, recovery=recovery, output_file=output_file,
    549                         execute_once=execute_once)
    550 # Interaction: interleave the interaction and non-interaction kernel for each time step.
    551 # E.g. Inter -> Normal -> Inter -> Normal if endtime-time == 2*dt
    552 else:
    553     cur_time = time

File ~\Anaconda3\envs\parcels\Lib\site-packages\parcels\kernel\kernelsoa.py:213, in KernelSOA.execute(self, pset, endtime, dt, recovery, output_file, execute_once)
    211 recovery_kernel = recovery_map[p.state]
    212 p.set_state(StateCode.Success)
--> 213 recovery_kernel(p, self.fieldset, p.time)
    214 if p.isComputed():
    215     p.reset_state()

File ~\Anaconda3\envs\parcels\Lib\site-packages\parcels\tools\statuscodes.py:182, in recovery_kernel_out_of_bounds(particle, fieldset, time)
    178 """Default sampling error kernel that throws OutOfBoundsError."""
    179 if particle.exception is None:
    180     # TODO: JIT does not yet provide the context that created
    181     # the exception. We need to pass that info back from C.
--> 182     raise OutOfBoundsError(particle, fieldset)
    183 else:
    184     error = particle.exception

OutOfBoundsError: 0
Particle P[14](lon=4.700000, lat=60.000000, depth=0.000000, time=0.000000)
Time: 2022-01-01T00:00:00.000000000
timestep dt: 3600.0
Out-of-bounds sampling by particle at (4.7, 60.0, 0.0)`

Any idea why this is happening?

**Comment by michaeldenes** [comment]
Hi @Xyx105,

I'm not entirely sure what is causing your error, could you paste in the result of:
`print(np.min(fieldset.U.lon), np.max(fieldset.U.lon), np.min(fieldset.U.lat), np.max(fieldset.U.lat))`

Additionally, you're loading in the 40th level of your depth, but your particle is initialised at depth 0m (so even if your code didn't throw an error, your particle is unlikely to move about).

Cheers,
Michael

**Reply by Xyx105** [reply]
Hi @michaeldenes 

Here is the output of that:
9.96921e+36 9.96921e+36 9.96921e+36 9.96921e+36

This doesn't seem normal. Let me know what you think.

Regards,
Jack

**Reply by michaeldenes** [reply]
Hi @Xyx105,
That is indeed quite odd! The longitudes/latitudes should have a fairly normal range (e.g. [-180,180) and [-90,90], respectively).
What does the grid of your velocity data look like? Could you share the full example of your code?

Thanks,
Michael

**Reply by Xyx105** [reply]
Hi @michaeldenes the first timestep on my grid can be found here: https://drive.google.com/file/d/1QNQ06ycDU0BQVCaAuiLMJ4AqetqVHa_m/view?usp=drive_link

My code looks like this`from parcels import( AdvectionRK4,
    FieldSet,
    JITParticle,
    ParticleFile,
    ParticleSet,
    AdvectionRK4_3D,
    ErrorCode,
                   )
from netCDF4 import Dataset
from datetime import timedelta
from glob import glob 
import numpy as np
from scipy.interpolate import interp2d, RectBivariateSpline
from datetime import timedelta as delta
from os import path
import glob 
import numpy as np
import netCDF4 as nc
import pandas as pd
import xarray as xr
import matplotlib.pyplot as plt






    

filenames = {
    "U": {
        "lon": f"output.nc",
        "lat": f"output.nc",
        "data": f"NC Files/NEATL*.nc",
    },
    "V": {
        "lon": f"output.nc",
        "lat": f"output.nc",
        "data": f"NC Files/NEATL*.nc",
    },
}
             

variables = {'U': 'u',
             'V': 'v'}
             


dimensions = {"lon": "lon_u", "lat": "lat_u", "time": "ocean_time"}

#indices = {'depth' : [40]}

fieldset = FieldSet.from_nemo(
    filenames, variables, dimensions, allow_time_extrapolation=True,
)


#indices = {'depth' : [40]}

fieldset.computeTimeChunk(fieldset.U.grid.time[0], 6)
fieldset.computeTimeChunk(fieldset.V.grid.time[0], 6)




npart = 1
lonp = 4.7 
latp = 60 



pset = ParticleSet.from_list(fieldset, JITParticle, lon=lonp, lat=latp,  )

print(np.min(fieldset.U.lon),
np.max(fieldset.U.lon),
np.min(fieldset.U.lat),
np.max(fieldset.U.lat))


pfile = ParticleFile("nemo_particles", pset, outputdt=delta(days=5))
kernels = pset.Kernel(AdvectionRK4) 

pset.show(fieldset.U)

pset.execute(kernels, runtime=delta(days=1), dt=delta(hours=6), output_file=pfile)`

**Reply by michaeldenes** [reply]
Hi @Xyx105, I've had a look at your dataset, but I get a different outcome using the following code snippet

```
import numpy as np
from parcels import FieldSet

variables = {'U': 'u', 'V': 'v'}
dimensions = {"lon": "lon_u", "lat": "lat_u", "time": "ocean_time"}
fieldset = FieldSet.from_nemo('data/NEATL*.nc', variables, dimensions, allow_time_extrapolation=True)

print(np.min(fieldset.U.lon), np.max(fieldset.U.lon), np.min(fieldset.U.lat), np.max(fieldset.U.lat))
```

Which produces:
` (-12.663201, 0.1604333, 49.012863, 56.72801) `

I'm not sure what's in your output.nc files, but maybe there is something wrong in there.


Additionally, if you have a look at where your data is, the position (4.7°E, 60°N) isn't inside your fieldset.
![image](https://github.com/OceanParcels/parcels/assets/1852917/bf528d5c-d9a8-473c-a6d9-b642dec87006)

If I place a particle inside that coloured box, I get the desired outcome.

Cheers,
Michael

**Reply by Caiyuan6** [reply]
Hi @Xyx105, may I ask if your question has been resolved? If so, how was it resolved?

**Comment by keli559** [comment]
I have the same error, out-of-bounds error. And code 
`from parcels import ErrorCode `
Does not work. It is telling me ErrorCode doesn't exist. I know this code was working last year.

**Reply by erikvansebille** [reply]
Yes indeed, since [Parcels v3.0](https://github.com/OceanParcels/parcels/releases/tag/v3.0.0), we have changed the way that Errors are dealt with. This is not backward-compatible, hence the error when you import `ErrorCode`.

You can now write a Kernel (see example below) and add that as a normal Kernel to your `pset.execute()`:

```python
def DeleteErrorParticle(particle, fieldset, time):
    if particle.state >= 40:  # deletes every particle that throws an error
        particle.delete()
```

See also [this part of the Kernel Loop tutorial](https://docs.oceanparcels.org/en/latest/examples/tutorial_kernelloop.html#Working-with-Status-Codes)
